ni_measurementlink_service/session_management/_reservation.py

```diff
@@ -94,6 +142,52 @@ def _grpc_channel_pool(self) -> GrpcChannelPool:
             raise ValueError("This method requires a gRPC channel pool.")
         return self._session_manager._grpc_channel_pool
 
+    @cached_property
+    def _reserved_pin_or_relay_names(self) -> Iterable[str]:
+        # Use dict keys to preserve insertion order.
+        names = {}
+        for session_info in self._session_info:
+            for channel_mapping in session_info.channel_mappings:
+                names[channel_mapping.pin_or_relay_name] = True
+        return list(names.keys())
+
+    @cached_property
+    def _reserved_sites(self) -> Iterable[int]:
+        # Use dict keys to preserve insertion order.
+        sites = {}
+        for session_info in self._session_info:
+            for channel_mapping in session_info.channel_mappings:
+                sites[channel_mapping.site] = True
+        return list(sites.keys())
+
+    @cached_property
+    def _reserved_instrument_type_ids(self) -> Iterable[str]:
+        # Use dict keys to preserve insertion order.
+        ids = {}
+        for session_info in self._session_info:
+            ids[session_info.instrument_type_id] = True
+        return list(ids.keys())
```

Comment:
So..., dictionary insertion is O(1),..., but so is list appending...

( note: that `dict.keys()` is O(n), so current implementation is O(n<sup>2</sup>)!! )

Which is simpler?
```suggestion
    @cached_property
    def _reserved_instrument_type_ids(self) -> Iterable[str]:
        # Use dict keys to preserve insertion order.
        ids = list()
        for session_info in self._session_info:
            ids.append(session_info.instrument_type_id)
        return ids
```

Reply:
Appending to a list doesn't remove duplicates.

Reply:
> ( note: that dict.keys() is O(n), so current implementation is O(n2)!! )

This is a cached property, so it only executes once.

Reply:
> Appending to a list doesn't remove duplicates.

If this is a reason to use a dict, then the comment should say this.

Reply:
Also, if you REALLY REALLY care about order -> which should be taken, the first occurrence of a name or the last? (what you currently have will be the first):


```
>python
Python 3.9.6 (tags/v3.9.6:db3ff76, Jun 28 2021, 15:26:21) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> data = {}
>>> for x in "apple_apple":
...     data[x] = True
...
>>> data
{'a': True, 'p': True, 'l': True, 'e': True, '_': True}
```

Reply:
Update: I am now using `dict_keys` as an ordered set. It supports O(n) iteration and O(1) `x in set` (amortized best case).

```python
# The `dict_view` type is a set-like type. In Python 3.7 and later, dictionaries
# preserve insertion order.
#
# Note: the set difference operator does not preserve order.
def _to_ordered_set(values: Iterable[_T]) -> AbstractSet[_T]:
    return dict.fromkeys(values).keys()
```

Reply:
> note: that dict.keys() is O(n)

BTW, in current versions of Python, `dict.keys()` returns a view object, which is O(1). Iterating over the keys is O(n).

Reply:
> BTW, in current versions of Python, dict.keys() creates a view object, which is O(1). Iterating over the keys is O(n).

This is really what I meant. Thank you for phrasing better.